### ISSUE TITLE
GuiEffectChecker has broken override checks for @PolyUI

### ISSUE THREAD

**Issue by csgordon** [typetools/checker-framework#1495]
As noticed in @lazaroclapp's PR #1481, the GuiEffectChecker has a bug that is roughly equivalent to unchecked covariant subtyping on generic parameters.

For a concrete example:
```java
import org.checkerframework.checker.guieffect.qual.*;

public class Q {

    @PolyUIType
    public static interface I {
        @PolyUIEffect
        public void m();
    }

    public static void reqSafe(@AlwaysSafe I i) {}
    @UIEffect
    public static void reqUI(@UI I i) {}

    @PolyUIType
    public static interface Doer {
        public void doStuff(@PolyUI Doer this, @PolyUI I i);
    }

    public void q(@AlwaysSafe Doer doer, @UI I i) {
        doer.doStuff(i);
    }

    public static void main(String[] args) {
        @AlwaysSafe Doer d = new @AlwaysSafe Doer() {
                                @SafeEffect
                                public void doStuff(@AlwaysSafe I i) {
                                    reqSafe(i);
                                }
                              };
        @UI I ui = new @UI I() {
                        @UIEffect
                        public void m() {
                            reqUI(null);
                        }
                    };
        Q q = new Q();
        q.q(d, ui);
    }
}
```
The ```main``` method above feeds an ```@UI I``` into ```reqSafe```.  Some part of the code above should be rejected.  The critical point of failure is that at the call to method ```doStuff```, the ```@AlwaysSafe Doer``` is upcast to ```@UI Doer```, and ```@UI``` is chosen as the instantiation of ```@PolyUI```, in order to permit the passing of the ```@UI I```.  Then inside the concrete implementation of ```doStuff```, the method is permitted to continue believing the receiver and argument are both ```@AlwaysSafe```.

Essentially, the choice of implicit effect variable instantiation is made once outside the class, but we then permit it to be changed when we use ```@PolyUI``` receiver annotations.

I initially wondered if the bug was specific to the GuiEffectChecker or inherent to the handling of receivers annotated with ```@PolymorphicQualifer``` qualifiers.  However, the tainting checker does not have the same error.  Consider:

```java
import org.checkerframework.checker.tainting.qual.*;

public class Tainting {

    public static void reqUntainted(@Untainted Object o) {
        System.out.println("Should be untainted: "+o.toString());
    }

     public static interface ITaint {
         public void m(@PolyTainted ITaint this, @PolyTainted ITaint other);
     }
     @Untainted
     public static class UTaint implements ITaint {
         public void m(@Untainted UTaint this, @Untainted ITaint other) {
             reqUntainted(other);
         }
     }
     @Tainted
     public static class TTaint implements ITaint {
         public void m(@PolyTainted TTaint this, @PolyTainted ITaint other) {
         }
         @Override
         public String toString() { return "<Tainted object>"; }
     }

     public static void main(String[] args) {
        @Untainted ITaint un = new @Untainted UTaint();
        @Tainted ITaint t = new @Tainted TTaint();
        un.m(t);
     }
}
```
When compiling this with the TaintingChecker:
```
$ javac -processor org.checkerframework.checker.tainting.TaintingChecker Tainting.java 
Tainting.java:15: error: [type.invalid] [@Tainted] may not be applied to the type "@Tainted UTaint"
     public static class UTaint implements ITaint {
                   ^
Tainting.java:16: error: [override.param.invalid] void m(@Untainted UTaint this, @Untainted ITaint p0) in Tainting.UTaint cannot override void m(@PolyTainted ITaint this, @PolyTainted ITaint p0) in Tainting.ITaint; attempting to use an incompatible parameter type
         public void m(@Untainted UTaint this, @Untainted ITaint other) {
                                                                 ^
  found   : @Untainted ITaint
  required: @PolyTainted ITaint
Tainting.java:16: error: [override.receiver.invalid] void m(@Untainted UTaint this, @Untainted ITaint p0) in Tainting.UTaint cannot override void m(@PolyTainted ITaint this, @PolyTainted ITaint p0) in Tainting.ITaint; attempting to use an incompatible receiver type
         public void m(@Untainted UTaint this, @Untainted ITaint other) {
                     ^
  found   : @Untainted UTaint
  required: @PolyTainted ITaint
3 errors
```

So some part of the GuiEffectChecker has broken override checking.

There are two main options:

1. Restrict declarations: If we disallow using ```@PolyUI``` parameters with ```@PolyUIEffect``` methods (except for the receiver), this will avoid conflating a class-level "effect variable" with a method-level one, as currently happens.  It might, however, break some existing code.  
2. Restrict subtyping: If a class uses the class of methods the first option would reject, instead allow it but prohibit coercing ```@AlwaysSafe``` instances of that class to ```@UI```.  This would break a different subset of possible code, and possibly also make the checker's behavior harder to explain to non-type-theorists.

Based on the SWT and AWT code I've seen, the first option wouldn't be a problem, making it preferable.  But we should ask @msridhar and @lazaroclapp as representatives of the most substantial use we're aware of at the moment, on very different UI infrastructure.

@lazaroclapp, is the problematic case (the case you noticed is permitted when it should be rejected) suggestive of actual code you've seen, or were you just being thorough working through the possibilities?  Since either solution requires detecting the same class of problematic methods, I can prototype the first one for you to try out if you're unsure.

**Comment by lazaroclapp** [comment]
@csgordon The test case in the PR doesn't appear as such in our code. But we do have a lot of cases with `o.foo(@UI C() {...})`, where `C` is a `@PolyUI` class/interface with a single `@PolyUIEffect` method.

Here, the type of o's class is usually polymorphic, although it is polymorphic on our own type qualifiers (namely: `@UIThread`, `@SafeThread` and `@PolyUIThread`). We have our own check that if `o` is instantiated `@SafeThread`, then it shouldn't accept something instantiated `@UI`.

I would possibly be ok with disallowing `@PolyUI` arguments with `@PolyUIEffect` methods, but only in the cases in which the receiver of the method that takes that argument is also `@PolyUI`. That should fix this issue as well, correct? Otherwise this would block a lot of the places we use guieffect checker at, and in general it would make it hard to use this checker with any form of stream API similar to RxJava. The second alternative should also work for us. 

Either way I'd love to be able to test any tentative patch for this on our setup before it lands on master. Thanks!

**Comment by csgordon** [comment]
Let's clarify.  

First, it sounds like you're partially implementing option 2 in a secondary analysis.  Any particular reason?  Because of concerns about this issue?

For option 1, here's a plausible class definition with some methods that would be accepted or rejected:
```java
@PolyUIType
public class OverrideTest {
    @PolyUIEffect
    public void polyMethod1() {
        // OK, because the only @PolyUI argument is the receiver
    }
    //:: error: (override.invalid.polymorphism)
    @PolyUIEffect
    public polyMethod2(@PolyUI Iface x) {
        // Rejected, because the receiver implicitly picks up @PolyUI,
        // conflating external choice of receiver parameter with a per-method effect variable
    }
    @UIEffect
    public void polyMethod3(@PolyUI Iface x) {
        // This could be made to work if the receiver was assumed @UI
        // here.
        // Probably also okay for @AlwaysSafe receivers, since that would restrict
        // the method from being invoked on @UI instances
    }
    @SafeEffect
    public void polyMethod4(@PolyUI Iface y) {
        // Same as above, with @UI receiver --- making it @AlwaysSafe this would 
    }
}
```
Do the restrictions above seem workable?

**Comment by lazaroclapp** [comment]
Basically, what our secondary analysis is concerned with is that we want to subscribe/pass `@UI` instances only on streams that are marked as going to be executed in the main thread (`@UIThread`). So is not really implementing option 2, is just that the actual check ends up looking fairly similar.

Here is an example stub method for this other checker, in case it helps (from an rxjava method):

```
@SubscribeEffects({0})
@PolyThread Observable<T> doAfterNext(@PolyUI Consumer<? super T> arg0) @PolyThread;
```

The point is here `@PolyUI` is allowed to be `@UI` (on arguments at the indexes listed in `@SubscribeEffects`) iff `@PolyThread` is `@UIThread`.

Either way, because of this separate type hierarchy, I think option 1 should also work for us, now that I think about it. Since I don't think we pass a `@PolyUI` type as an argument to a method where the receiver is also `@PolyUIType` (we might pass it to a receiver which has our own `@PolyThread` annotation, but that should not be a problem as far as guieffect itself is concerned). I am trying to think if there is anything like `polyMethod2` in our scenario, but I can't think of a case.

**Comment by csgordon** [comment]
> The point is here @PolyUI is allowed to be @UI (on arguments at the indexes listed in @SubscribeEffects) iff @PolyThread is @UIThread.

Modulo not being filled in on ```@SubscribeEffects```, this sounds like an unsound approximation of what I called option 2 above.  Like option 2, you're making a restriction (a different one) at call sites (whereas option 2 imposes restrictions other places as well).  The intent of both is to ensure that if you're passing an ```@UI``` callback to a ```@PolyUI``` parameter of a method on class X, the instance of X really is a UI instance.

The challenge that my option 2 addresses is other routes beyond the immediate call site that would coerce a safe instance to a UI instance.  The repro code in the initial report relies on the receiver being upcast to ```@UI``` at the call site.  Imposing a restriction similar to what you describe, where we check if the receiver's type really is ```@UI``` --- and not a subtype thereof --- when ```@PolyUI``` is instantiated to ```@UI``` would reject the repro code, but not the following:
```java
...
@UI I i = /* as above */;
@AlwaysSafe Doer d = /* as above */;
@UI Doer danger = d; // permitted because @AlwaysSafe <: @UI
danger.doStuff(i);
```
On the last line, the static type of ```danger``` without any callsite subtyping is already ```@UI```, so a check specific to call sites of polymorphic methods is insufficient --- it would need to be known if any ```@AlwaysSafe``` receiver flowed to that callsite.  Restricting subtyping as in option 2 would reject both the original repro code, and this sketch.

I don't know the details of your approach, but if you have an ```@AnyThread``` or similar which is a subtype of ```@UIThread```, your system would also miss the version sketched above.  If your ```@UIThread``` has no subtypes (subqualifiers), then the check you describe is likely enough.

**Comment by csgordon** [comment]
I'll start prototyping option 1 to see if it works for you.

**Comment by csgordon** [comment]
@lazaroclapp @msridhar the issue1495 branch is the first cut at a fix.

I ended up not implementing option one or option 2, but instead more carefully reused the existing override support, which was nicely engineered for me to be just a bit more permissive on receiver qualifier overrides.  So rather than restricting subtyping or the signature of polymorphic methods, instead the typing of the body is restricted.  An override of a ```@PolyUIEffect``` method may replace the receiver with ```@AlwaysSafe Subclass``` if it is overriding a method inherited from ```@AlwaysSafe ParentClass```.  This is a much more precise restriction than the other options I proposed previously, and should reject less code.

It's possible you'll see new errors in your codebase using this patch, related to method overrides.  This is because the override checking on parameters was previously broken, but should now be fixed.  So if some bugs in your codebase fell through that hole, they should now be caught along with the issues from confusing polymorphic qualifiers.

**Comment by mernst** [comment]
@lazaroclapp @msridhar Does the issue1495 branch address your concerns?  Or have you noticed that other changes need to be made to that code?

**Comment by lazaroclapp** [comment]
@csgordon @mernst Sorry for the delayed response, the original issue branch somehow slipped from my radar. We cannot easily test against that branch exactly, but I just cherry-picked  `a8b32480c49dac14246765ddea1354c819f10332` on top of our internal CF fork (separate from the NullAway visible fork) and I found no problems with it in our code that were not real issues before (and trivial to fix).

We did find an instance which looked like (simplified, renamed):

```
@PolyUIType interface OurUICallback {
  @PolyUIEffect void onEvent(@PolyUI SomeT someObject);
}
```

There it is ambiguous whether the type of the receiver or the `someObject` parameter should determine the effect of `onEvent`. I believe our expectation was that either all three would be `@UI` or all three would be `@Safe`. However, this example was not an issue for us, since there is actually no use for a `@Safe` instance of that callback, and we could annotate it instead as:


```
interface OurUICallback {
  @UIEffect void onEvent(@UI SomeT someObject);
}
```

Let me know if there are more changes you are concerned about related to this issue, besides those in that particular commit. And again, thanks and sorry for the delay!

cc: @msridhar
